### DISCUSSION TITLE
Prune usage?

### DISCUSSION THREAD

**Discussion by dachtera** [SAP-archive/btp-setup-automator#318] [Q&A] [answered]
Hello - Is there anyone with a working syntax for -prunesubaccount?   

I have tried to follow the documentation, and instead of removing, btpsa tries to create the subaccount, not remove it.

ex:

./btpsa -prunesubaccount [sub account ID] -loginmethod sso -myemail [...] -globalaccount [my global account id] -region [myregion]

Thanks in advance.

**Comment by lechnerc77** [comment, marked as answer]
Hi,

the flow of the btpsa is:

- Try to create the resources s described in the `parameters.json` and `usecase.json` file
- Once everything is available (= the state defined in the above-mentioned files), the setup autromator:
  - if the parameter `pruneusecase` is set to true all resources as defined in the `usecase.json` will be deleted. The subaccount will not be deleted
  - if the parameter `prunesubaccount` is set to true all resources as defined in the `usecase.json` will be deleted. After that the subaccount will be removed
 
If the specified subaccount is not existing, btpsa will try to create it. 

Some points to consider: 

- The `prunesubaccount` parameter is a boolean, so in order to trigger the deletion you must specify `./btpsa -prunesubaccount true` to trigger it. The subaccount is specified via `-subaccountid [ID of subaccount]`
- if you do not specify the `usecase.json` and `parameters.json` file in the command line it will take the ones specified in your usecase, so all the information you do not provide via command line parameters will be filled up with the information from these files
- As there is no recursive delete make sure that the `usecase.json` contains all the services that are existing in the subaccount. They need to be deleted before the subaccount can be removed.

**Reply by dachtera** [reply]
Thanks - I believe I have this setup correctly based on what I understand.  So, First, I've created the sub-account, and when it's no longer needed, I want to remove everything, and use the following command (close to the same as creation statement without specifying the region and log file:

./btpsa -loginmethod sso -myemail [..] -parameterfile ./parameters-thin.json -globalaccount [...]-subaccountid [...] -prunesubaccount true -subaccountname timtest

When I run this, (with the correct global and sub IDs) it starts to create the items again and fails out when trying to recreate the same services already in use (because it's trying to allocate the same quantity of resource values, exceeding available amount)

Is there simply an option to provide the global and sub account ID to simply wipe out anything in that sub-account ignoring what was created originally?  I ask, as many times our developers will add in custom entitlements above the baseline configuration, which would also render the prunesubaccount process unusable as the configuration would have changed in various aspects from initial creation from the usecase file.

Removing all the options from my usecase, runs farther, but ends up with this error:

####################################################################################################
# Remove assets created for this use case
####################################################################################################
[2022-11-17 17:15:34] INFO       : Delete sub account
[2022-11-17 17:15:34] RUN COMMAND: btp --format json delete accounts/subaccount 'c2a6b2e6-4bac-43fe-b782-1a4518c5bfdd' --global-account '69434341-cb97-4d44-ad44-c44f4ab00e7c' --confirm
[2022-11-17 17:15:37] ERROR      : 
[2022-11-17 17:15:37] ERROR      : {
  "error": "You can't delete subaccounts with active application subscriptions and service instances. Your active resources in this subaccount are: Service instances / Application subscriptions. Remove these resources first, then try again. [Error: 70010/409]"
}

FAILED

I was of the understanding that the prune function remove *.* from under the sub-account, including any active subscriptions or services.  Is this not the case?

**Comment by lechnerc77** [comment]
There is no option to "just" delete a subaccount with btpsa. There is also no such option in the BTP CLI. As I wrote in my response if the subaccount contains app subscriptions or service instance it cannot be deleted (Thats where your error comes from). They need to be cleaned up first. That's why they need to be specified in the usecase file that is then used to delete these resources, wait until deletion is done and then deletes the subaccount. 

I cannot reproduce the behavior you described in a regulat BTP account. If you think you run into a bug, you can open an issue. Please attach the logs, the parameters.json and the usecase.json you used and make sure that the data is anonymized where necessary.

**Reply by dachtera** [reply]
Thanks - I'm good.  I sounds as if the tool will not work the way our team would need it to, and current presents no increase in efficiency vs. manually removing sub-accounts from the UI.  Thank you for your feedback!
